main.cpp

```diff
@@ -51,22 +52,58 @@ const char *host_name = "echo.mbedcloudtesting.com";
 // Echo server port (same for TCP and UDP)
 const int port = 7;
 
-Mutex PrintMutex;
+static rtos::Mutex trace_mutex;
+
+#if MBED_CONF_MBED_TRACE_ENABLE
+static void trace_wait()
+{
+    trace_mutex.lock();
+}
+
+static void trace_release()
+{
+    trace_mutex.unlock();
+}
+
+
+static uint32_t cellular_starttime = us_ticker_read() / 1000L;
+static char time_st[50];
+
+static char* trace_time(size_t ss)
+{
+    snprintf(time_st, 49, "[%08lums]", ((us_ticker_read()-cellular_starttime) / 1000L));
+    return time_st;
+}
+
+static void trace_open()
+{
+    mbed_trace_init();
+    mbed_trace_prefix_function_set( &trace_time );
+
+    mbed_trace_mutex_wait_function_set(trace_wait);
+    mbed_trace_mutex_release_function_set(trace_release);
+}
+
+static void trace_close()
+{
+    mbed_trace_free();
+}
+#endif // #if MBED_CONF_MBED_TRACE_ENABLE
+
 Thread dot_thread(osPriorityNormal, 512);
 
 #define PRINT_TEXT_LENGTH 128
 char print_text[PRINT_TEXT_LENGTH];
 void print_function(const char *input_string)
 {
-    PrintMutex.lock();
+    trace_mutex.lock();
     printf("%s", input_string);
     fflush(NULL);
-    PrintMutex.unlock();
+    trace_mutex.unlock();
```

Comment:
fflush could be removed

Reply:
removed

Reply:
You should only need the flush if you're not printing a newline, as newline is supposed to flush. Always fflushing for everything is going to be a bit clunky. Maybe you could have a speparate case just for the dot. And you presumably mean `fflush(stdout)` anyway?

What's the print_text buffer for? Why aren't you just using `vprintf`?

Reply:
I have no idea why there was print_text buffer. Changed to use vprintf.
Also now only dot printing thread uses fflush.